src/markdown_exec/formatters/typescript.py

```diff
@@ -0,0 +1,197 @@
+            )
+            check_proc.wait()
+            stderr = check_proc.stderr.read()
+            if check_proc.returncode != 0:
+                raise ExecutionError(code_block("typescript", stderr, **extra))
+        finally:
+            os.remove(file.name)
+
+
+def _get_session_data(session: str | None, code: str) -> SessionData:
+    if session is None:
+        return {
+            "code": code.split("\n"),
+        }
+
+    if session not in _sessions:
+        _sessions[session] = {
+            "code": [],
+        }
+
+    _sessions[session]["code"].extend(code.split("\n"))
+
+    return _sessions[session]
+
+
+def _run_typescript(
+    code: str,
+    returncode: int | None = None,
+    session: str | None = None,
+    id: str | None = None,
+    **extra: str,
+) -> str:
+    _exec_lock.acquire(blocking=True)
+
+    try:
+        session_data = _get_session_data(session, code)
+
+        _check_types(session_data, **extra)
+
+        try:
+            req = ExecRequest(session=session, code=code.split("\n"))
+            if session is None:
+                del req["session"]
+            req_json = json.dumps(req) + "\n"
+            print(f"writing request: {req_json}")
+            proc.stdin.write(req_json)
+            proc.stdin.flush()
+
+            start_time = time.time()
+
+            while proc.poll() == None and time.time() - start_time < _TIMEOUT:
+                try:
+                    line = exec_output.get()
+                    print(f"got line: {line}")
+                    break
+                except queue.Empty:
+                    time.sleep(0.1)
+
+            if proc.poll() is not None:
+                print("process died")
+                raise ExecutionError("Deno process died unexpectedly")
+
+            if time.time() - start_time >= _TIMEOUT:
+                print("timeout")
+                raise ExecutionError("TypeScript execution timed out")
+
+            result: ExecResponse = json.loads(line)
+
+            if result["result"] == "error":
+                error_lines = [
+                    entry["message"]
+                    for entry in result["output"]
+                    if entry["type"] == "stderr"
+                ]
+                print(f"error: {os.linesep.join(error_lines)}")
+                raise ExecutionError(
+                    code_block(
+                        "typescript",
+                        "\n".join(error_lines),
+                        **extra,
+                    )
+                )
+            else:
+                output = "\n".join(
+                    [
+                        entry["message"]
+                        for entry in result["output"]
+                        if entry["type"] == "stdout"
+                    ]
+                )
+                return output
+
+        except BrokenPipeError:
+            raise ExecutionError("Deno process died unexpectedly")
+    finally:
+        _exec_lock.release()
```

Comment:
lint: I'd recommend avoiding such large `try` blocks. Maybe split out the inner logic into separate functions. Ideally the `try` blocks have only one line.

Reply:
yeah, this was a result of just (ahem) trying things to understand where my race condition is coming from - I was originally using the context manager protocol on `Lock` for this (so it was a `with` block rather than a `try`/`finally`), but at one stage I was finding that things were processing out of order so I decided to make the lock/unlock more explicit.

worst case if I need to keep the try/finally I'll break this down into internal functions with more readable names and then call them from the `try` block.